### ISSUE TITLE
Upload security-nonce not being found

### ISSUE THREAD

**Issue by EtienneBruines** [squell/bb-scripts#13]
Not entirely sure whats wrong, but after the update, the `upload.sh` script no longer fetches the correct nonce. 

Debugging now. 

---
## Update 1
Apparently 
```bash
sed -n '/<form/,${/nonce/s/^.*name=.\([.[:alpha:]]\+\).*value=.\([0-9a-f-]\+\).*$/\1=\2/p;q}'
``` 
returns doesn't filter the nonce correctly; for some reason :confused: 

## Update 2
The line it should parse, looks something like this (and is indeed present):
```html
<form name="uploadGradebookForm2" method="post" action="/webapps/gradebook/do/instructor/uploadGradebook2?course_id=_107298_1" enctype="multipart/form-data" onsubmit="return validateForm(this);"><input type="hidden" name="course_id" value="_107298_1"><input type="hidden" name="actionType" value="processFile"><input type='hidden' name='blackboard.platform.security.NonceUtil.nonce' value='b04249d6-ccad-49c7-9bf2-7ca8f2cd2e2e3'>
```
(the *only* line in the file where there's no newline before that `input` field, and therefore the `sed` script can do its job)

**Comment by squell** [comment]
Similar problem occurs on the download-side, debugging that atm as well.

**Comment by squell** [comment]
In getsch.sh, the issue was that there are multiple forms in the html now; locating the correct form and replacing "<form" with something more appropriate (e.g. "mainForm" or something) might help locate the correct nonce.

**Comment by EtienneBruines** [comment]
Wow, did not expect that. (I'm no `sed` expert).

Guess you're right, it had trouble finding the correct one (making it print none at all).

For the `upload.sh` file, the `sed` command would become:
```bash
sed -n '/uploadGradebookForm2/,${/nonce/s/^.*name=.\([.[:alpha:]]\+\).*value=.\([0-9a-f-]\+\).*$/\1=\2/p;q}'
```

**Comment by squell** [comment]
Does this fix upload.sh? I'm having trouble testing it.

**Comment by EtienneBruines** [comment]
It works at my computer; I can only hope it works on others.  ("only" have 1 Blackboard account to test it with, and there it works)

**Comment by EtienneBruines** [comment]
```
    <table class="bouncer" summary="Top frame table">
      <tr>
        <td>
          <div class="topTabs bgBanner" id="topTabs">
  
            <div class="brandingImgWrap">
              <a href="http://blackboard.ru.nl/webapps/portal/frameset.jsp" target="_blank" title="Radboud Universiteit Nijmegen">
                <img  src="/branding/_1_1/logoRU.png"  title="Radboud Universiteit Nijmegen" alt="Radboud Universiteit Nijmegen" class="bannerImage"/>
</a>
            </div>
  
            <div class="tabWrapper-left">
              <H2 CLASS="hideoff">Top Frame Tabs</H2>
              <table class="appTabs transparent" id="appTabList" summary="Tab List table">
                <tr>
                 <td id="My Blackboard"  >
                      <a href="/webapps/portal/execute/tabs/tabAction?tab_tab_group_id=_141_1" TARGET="_top"><span>My Blackboard</span>
                        <span class="hideoff"> Tab 1 of 4</span>
                      </a>
                    </td>
                  <td id="Courses.label"  class="active"  >
                      <a href="/webapps/portal/execute/tabs/tabAction?tab_tab_group_id=_3_1" TARGET="_top"><span>Courses</span>
                        <span class="hideoff"> Tab 2 of 4&nbsp;(active tab)</span>
                      </a>
                    </td>
                  <td id="Organisations"  >
                      <a href="/webapps/portal/execute/tabs/tabAction?tab_tab_group_id=_4_1" TARGET="_top"><span>Organisations</span>
                        <span class="hideoff"> Tab 3 of 4</span>
                      </a>
                    </td>
                  <td id="NWI"  >
                      <a href="/webapps/portal/execute/tabs/tabAction?tab_tab_group_id=_20_1" TARGET="_top"><span>NWI</span>
                        <span class="hideoff"> Tab 4 of 4</span>
                      </a>
                    </td>
                  </tr>
              </table>
```

When putting this through `html2text`, it becomes blank.

**Comment by squell** [comment]
That was the problem I'm referring too. Not sure *why*; adding <html><body> </body> etc doesn't work.

**Comment by EtienneBruines** [comment]
It's malformed, in the sense that there are no closing tags for the 'outer' table.

**Comment by squell** [comment]
Ah. Easy to fix that then.

**Comment by EtienneBruines** [comment]
Ah, there is no preview table to be found. 

```
Error: The following columns have invalid Column IDs: Opdracht 4
Error: The following usernames are invalid: s4264738, s4264738, s4264738
```

**Comment by squell** [comment]
This should fix the preview: 364408cfc05520cd33c7e3d52ed656c0505dd22a

**Comment by squell** [comment]
The only problem remaining for me is that upload.sh doesn't work if bblogin2.sh thinks I'm already logged in, which is strange.

**Comment by EtienneBruines** [comment]
Did not have [that](https://github.com/squell/bb-scripts/issues/13#issuecomment-170636463) problem without 364408cfc05520cd33c7e3d52ed656c0505dd22a (but had to log-in each time, and every 2nd attempt failed)

**Comment by squell** [comment]
I think the problem can be fixed by replacing wget with curl everywhere. Which is a change which is already in the experimental branch. Since things are breaking atm anyway, a good moment to apply those changes to 'master'.

**Comment by EtienneBruines** [comment]
Hmm, right now having the issue of Access Denied ... 

`blackboard.platform.security.NonceUtil.nonce` is being shortened to `blackboard` for some reason.

**Comment by squell** [comment]
At what stage in the script? Just uploaded a grade successfully in an old course I still have access to.

**Comment by EtienneBruines** [comment]
Well, the `$NONCE` variable has this value: `blackboard=776512b1-40e1-4a50-b5e6-201fb9612d2e`

**Comment by squell** [comment]
I think this issue can be closed then? Great job on the debugging session. :)

**Comment by EtienneBruines** [comment]
Part of the code was using curl, the other part was using wget... That didn't go too well. Using your fix at 364408cfc05520cd33c7e3d52ed656c0505dd22a works :-). 

You too :)
